packages/cozy-procedures/src/redux/documentsDataSlice.js

```diff
@@ -0,0 +1,121 @@
+import get from 'lodash/get'
+import creditApplicationTemplate from '../templates/creditApplicationTemplate'
+
+const documentsSlice = createSlice({
+  initialState: {
+    data: {
+      /*  payslip: {
+        ...files
+      } */
+    },
+    ui: {
+      /* payslip: {
+        loading: true,
+        error: false
+      } */
+    }
+  },
+  slice: 'documents',
+  reducers: {
+    init: (state, action) => {
+      return {
+        data: Object.keys(action.payload).reduce((acc, fieldId) => {
+          acc[fieldId] = {}
+          return acc
+        }, {}),
+        ui: Object.keys(action.payload).reduce((acc, fieldId) => {
+          acc[fieldId] = {
+            loading: false,
+            error: false
+          }
+          return acc
+        }, {})
+      }
+    },
+    update: (state, action) => {
+      console.warn('update documents slice not implemented')
+      return state
+    },
+    fetchDocumentLoading: (state, action) => {
+      const { idDoctemplate, loading } = action.payload
+      state.ui[idDoctemplate].loading = loading
+    },
+    fetchDocumentSuccess: (state, action) => {
+      const { idDoctemplate, loading, file } = action.payload
+      state.ui[idDoctemplate].loading = loading
+      state.data[idDoctemplate].document = file
+    },
+    fetchDocumentError: (state, action) => {
+      const { idDoctemplate, error } = action.payload
+      state.ui[idDoctemplate].error = error
+    }
+  }
+})
+const selectors = {
+  getSlice: state => get(state, documentsSlice.slice)
+}
+
+const { actions, reducer } = documentsSlice
+export const {
+  init,
+  update,
+  fetchDocumentLoading,
+  fetchDocumentSuccess,
+  fetchDocumentError
+} = actions
+
+export function fetchDocument(client, documentTemplate) {
+  return async dispatch => {
+    dispatch(
+      fetchDocumentLoading({
+        loading: true,
+        idDoctemplate: documentTemplate
+      })
+    )
+    try {
+      /* const docWithRules =
+          creditApplicationTemplate.documents[documentTemplate]
+        
+        //Ajout du metadata.
+        // Ajout du order by
+        const cozyRules = {
+          trashed: false,
+          type: 'file',
+          ...docWithRules.rules
+        }
+        const files = await client.collection('io.cozy.files').find(cozyRules)
+
+        console.log('files', files) */
+
+      const files = {}
+      if (files.data) {
+        dispatch(
+          fetchDocumentSuccess({
+            loading: false,
+            file: files.data[0],
+            idDoctemplate: documentTemplate
+          })
+        )
+      }
+    } catch (error) {
+      dispatch(
+        fetchDocumentError({
+          error: error.message,
+          idDoctemplate: documentTemplate
+        })
+      )
+    }
+    dispatch(
+      fetchDocumentLoading({
+        loading: false,
+        idDoctemplate: documentTemplate
+      })
+    )
+  }
+}
+
+export const { getSlice } = selectors
+export default reducer
```

Comment:
I don't understand why we need a reducer/slice for this ? cozy-client already has a redux store managing documents.

Reply:
Discussed with @Crash-- via zoom. Since some of the query operations are not supported directly via cozy-client, they need to be implemented downstream and thus need downstream error management. It would be interesting to see if those downstream filtering of documents could be a functionality of cozy-client so that queries directly support client-side filtering for example (and having errors automatically managed for example if a document is not found).

It seems to me though that the loading could be implemented as a selector on cozy-client's state. I am not sure though if we want this since this would mean sharing cozy-client internal structure with the app state.

Reply:
I think that `cozy-client`'s state and this state does not represent the same thing. In `cozy-client`, we store the fetched documents whereas here we need to store the documents that are part of the credit application. Maybe that technically you could mark the documents that are part of the procedure in `cozy-client`'s state but I don't like the idea of using the same state for both.